### ISSUE TITLE
Question about Multi-turn Dialogue Generation based on GPT-2 Model

### ISSUE THREAD

**Issue by honghonghonghonghonghonghonghong** [facebookresearch/ParlAI#4460]
Hi there,

I am wondering that assume we have a single dialogue session such as Sentence A, Sentence B, Sentence C, Sentence D. Does ParlAI create three  instances AB, ABC, ABCD during training, for GPT-based models? Or it is just a single instance ABCD?

Thank you very much. 

Best

**Comment by klshuster** [comment]
This depends on how you set up your task - it is entirely up to the user

**Comment by honghonghonghonghonghonghonghong** [comment]
> 

Thank you for the reply.

What setup do you exactly refer to?

For example, if I run using the default setup without changing any arguments on ConvAI with GPT-2, it seems that the number of training steps reflects the fact that ParlAI creates 3 instances described above, i.e., AB, ABC, ABCD.

If I may, I would like to ask another question... In this case, how is the loss calculated? That is, for the instance ABC, is the loss for BC calculated, or only the last utterance is countered for loss calculation?

Thank you very much!

**Comment by klshuster** [comment]
Yes so it depends on the task, indeed that is how it works for ConvAI2; specifically, a conversation of utterances ABCDEF in that dataset will be split up into two conversations with two sets of targets

_Context_ A, _Target_ B
_Context_ ABC, _Target_ D
_Context_ ABCDE, _Target_ F
And
_Context_ '', _Target_ A
_Context_ AB, _Target_ C
_Context_ ABCD, _Target_ E

> In this case, with the default setup, how is the loss calculated for the GPT-based autoregressive model? That is, for the instance ABC, is the loss for both words in B and C are calculated, or is only the last utterance, C, countered for loss calculation?

The above explanation might answer this, but to abundantly clear --> given a context AB, and target C, the loss is over the target C tokens.

**Comment by honghonghonghonghonghonghonghong** [comment]
> Yes so it depends on the task, indeed that is how it works for ConvAI2; specifically, a conversation of utterances ABCDEF in that dataset will be split up into two conversations with two sets of targets
> 
> _Context_ A, _Target_ B _Context_ ABC, _Target_ D _Context_ ABCDE, _Target_ F And _Context_ '', _Target_ A _Context_ AB, _Target_ C _Context_ ABCD, _Target_ E
> 
> > In this case, with the default setup, how is the loss calculated for the GPT-based autoregressive model? That is, for the instance ABC, is the loss for both words in B and C are calculated, or is only the last utterance, C, countered for loss calculation?
> 
> The above explanation might answer this, but to abundantly clear --> given a context AB, and target C, the loss is over the target C tokens.

Thank you very much for the abundantly clear answer! Is both two sets of targets used for training for ConvAI2, or only the agent turn is used?

**Comment by klshuster** [comment]
Unless I am mistaken, in the ConvAI2 setup, we display *both* sides of the conversation, but they are presented *as separate training examples*; so, both sets of targets are used, but not at the same time.
